mantidproject/mantid
# add Placzek correction algorithm

This PR adds an algorithm `CalculatePlaczekSelfScattering` that generates the Placzek correction from
a workspace containing an incident spectrum and it's derivative such as is generated by `FitIncidentspectrum`. The `InputWorkspace` must have detector info and a material sample.

This PR also makes a change to `FitIncidentSpectrum` so that it keeps the detector data from it's input workspace.

**To test:**

1. Check that `CalculatePlaczekSelfScattering` returns a sane result, i.e. that the correction is actualy what is expected from the Placzek correction equation.

The following script runs the algorithm for some dummy data, but it will need to be tested with some real data.

```
from __future__ import (absolute_import, division, print_function, unicode_literals)

# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *

import matplotlib.pyplot as plt

import numpy as np

# Create the workspace to hold the already corrected incident spectrum
incident_wksp_name = 'incident_spectrum_wksp'
binning_incident = "%s,%s,%s" % (0.1, 0.02, 5.0)
binning_for_calc = "%s,%s,%s" % (0.2, 0.02, 4.0)
binning_for_fit = "%s,%s,%s" % (0.15, 0.02, 4.5)
incident_wksp = CreateSampleWorkspace(
    OutputWorkspace=incident_wksp_name,
    NumBanks=10,
    PixelSpacing=0.1,
    XMin=0.1,
    XMax=5.0,
    BinWidth=0.002,
    Xunit='Wavelength')
incident_wksp = ConvertToPointData(InputWorkspace=incident_wksp)

# Spectrum function given in Milder et al. Eq (5)
def incidentSpectrum(wavelengths, phiMax, phiEpi, alpha, lambda1, lambda2,
                     lamdaT):
    deltaTerm = 1. / (1. + np.exp((wavelengths - lambda1) / lambda2))
    term1 = phiMax * (
        lambdaT**4. / wavelengths**5.) * np.exp(-(lambdaT / wavelengths)**2.)
    term2 = phiEpi * deltaTerm / (wavelengths**(1 + 2 * alpha))
    return term1 + term2

# Variables for polyethlyene moderator at 300K
phiMax = 6324
phiEpi = 786
alpha = 0.099
lambda1 = 0.67143
lambda2 = 0.06075
lambdaT = 1.58

# Add the incident spectrum to the workspace
corrected_spectrum = incidentSpectrum(
    incident_wksp.readX(0), phiMax, phiEpi, alpha, lambda1, lambda2, lambdaT)
incident_wksp.setY(0, corrected_spectrum)

incident_spectrum = FitIncidentSpectrum(
    InputWorkspace='incident_wksp', 
    OutputWorkspace='fit_wksp', 
    BinningForCalc='0.2,0.01,4', 
    BinningForFit='0.2,0.01,4')
SetSampleMaterial(
    InputWorkspace='fit_wksp',
    ChemicalFormula='Co')
CalculatePlaczekSelfScattering(
    InputWorkspace='fit_wksp', 
    OutputWorkspace='placzek_scattering')
```

Fixes #26688 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
